health/healthService.go

```diff
@@ -0,0 +1,141 @@
+package health
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"runtime"
+	"sync"
+	"time"
+
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/core"
+)
+
+var log = logger.GetOrCreate("health")
+
+type healthService struct {
+	config                     config.HealthServiceConfig
+	folder                     string
+	cancelFunction             func()
+	records                    *records
+	diagnosableComponents      []diagnosable
+	diagnosableComponentsMutex sync.RWMutex
+}
+
+// NewHealthService creates a new HealthService object
+func NewHealthService(config config.HealthServiceConfig, workingDir string) *healthService {
+	log.Info("NewHealthService", "config", config)
+
+	folder := path.Join(workingDir, config.FolderPath)
+	records := newRecords(config.NumMemoryRecordsToKeep)
+
+	return &healthService{
+		config:                config,
+		folder:                folder,
+		cancelFunction:        func() {},
+		records:               records,
+		diagnosableComponents: make([]diagnosable, 0),
+	}
+}
+
+// RegisterComponent registers a diagnosable component
+func (h *healthService) RegisterComponent(component interface{}) {
+	h.diagnosableComponentsMutex.Lock()
+	defer h.diagnosableComponentsMutex.Unlock()
+
+	asDiagnosable, ok := component.(diagnosable)
+	if !ok {
+		log.Error("healthService.RegisterComponent(): not diagnosable", "component", fmt.Sprintf("%T", component))
+		return
+	}
+
+	h.diagnosableComponents = append(h.diagnosableComponents, asDiagnosable)
+}
+
+// Start starts the health service
+func (h *healthService) Start() {
+	log.Info("healthService.Start()")
+
+	h.prepareFolder()
+	ctx := h.setupCancellation()
+	go h.monitorContinuously(ctx)
+}
+
+func (h *healthService) prepareFolder() {
+	err := os.MkdirAll(h.folder, os.ModePerm)
+	if err != nil {
+		log.Error("healthService.prepareFolder", "err", err)
+	}
+}
+
+func (h *healthService) setupCancellation() context.Context {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	h.cancelFunction = cancelFunc
+	return ctx
+}
+
+func (h *healthService) monitorContinuously(ctx context.Context) {
+	for i := 0; h.shouldContinueInfiniteLoop(ctx); i++ {
+		shouldMonitorMemory := i%h.config.IntervalVerifyMemoryInSeconds == 0
```

Comment:
this function can be made more go idiomatic, and we can have other intervals (not necessarily multiples of 1 second):
```GO
    chanMonitorMemory := time.After(IntervalVerifyMemoryInSeconds * time.Second)
    chanDiagnoseComponents := time.After(IntervalDiagnoseComponentsInSeconds * time.Second)
    chanDiagnoseComponentsDeeply := time.After(IntervalDiagnoseComponentsDeeplyInSeconds * time.Second) 
    
    for{
        select{
        case <-chanMonitorMemory:
            h.monitorMemory()
            chanMonitorMemory := time.After(IntervalVerifyMemoryInSeconds * time.Second)
        case <-chanDiagnoseComponents:
            h.diagnoseComponents(false)
            chanDiagnoseComponents := time.After(IntervalDiagnoseComponentsInSeconds * time.Second)
        case <-chanDiagnoseComponentsDeeply:
            h.diagnoseComponents(true)
            chanDiagnoseComponentsDeeply := time.After(IntervalDiagnoseComponentsDeeplyInSeconds * time.Second)
        case <-ctx.Done():
            log.Info("healthService.monitorContinuously() ended")
            return    
        }
    }
```

Reply:
Changed.